contracts/PBAB+Collabs/GenArt721CoreV2WithEvents.sol

```diff
@@ -0,0 +1,798 @@
+
+    //All financial functions are stripped from struct for visibility
+    mapping(uint256 => address payable) public projectIdToArtistAddress;
+    mapping(uint256 => string) public projectIdToCurrencySymbol;
+    mapping(uint256 => address) public projectIdToCurrencyAddress;
+    mapping(uint256 => uint256) public projectIdToPricePerTokenInWei;
+    mapping(uint256 => address payable) public projectIdToAdditionalPayee;
+    mapping(uint256 => uint256) public projectIdToAdditionalPayeePercentage;
+    mapping(uint256 => uint256)
+        public projectIdToSecondaryMarketRoyaltyPercentage;
+
+    address payable public renderProviderAddress;
+    /// Percentage of mint revenue allocated to render provider
+    uint256 public renderProviderPercentage = 10;
+
+    mapping(uint256 => uint256) public tokenIdToProjectId;
+    mapping(uint256 => bytes32) public tokenIdToHash;
+    mapping(bytes32 => uint256) public hashToTokenId;
+
+    /// admin for contract
+    address public admin;
+    /// true if address is whitelisted
+    mapping(address => bool) public isWhitelisted;
+    /// true if minter is whitelisted
+    mapping(address => bool) public isMintWhitelisted;
+
+    /// next project ID to be created
+    uint256 public nextProjectId = 0;
+
+    modifier onlyValidTokenId(uint256 _tokenId) {
+        require(_exists(_tokenId), "Token ID does not exist");
+        _;
+    }
+
+    modifier onlyUnlocked(uint256 _projectId) {
+        require(!projects[_projectId].locked, "Only if unlocked");
+        _;
+    }
+
+    modifier onlyArtist(uint256 _projectId) {
+        require(
+            msg.sender == projectIdToArtistAddress[_projectId],
+            "Only artist"
+        );
+        _;
+    }
+
+    modifier onlyAdmin() {
+        require(msg.sender == admin, "Only admin");
+        _;
+    }
+
+    modifier onlyWhitelisted() {
+        require(isWhitelisted[msg.sender], "Only whitelisted");
+        _;
+    }
+
+    modifier onlyArtistOrWhitelisted(uint256 _projectId) {
+        require(
+            isWhitelisted[msg.sender] ||
+                msg.sender == projectIdToArtistAddress[_projectId],
+            "Only artist or whitelisted"
+        );
+        _;
+    }
+
+    /**
+     * @notice Initializes contract.
+     * @param _tokenName Name of token.
+     * @param _tokenSymbol Token symbol.
+     * @param _randomizerContract Randomizer contract.
+     */
+    constructor(
+        string memory _tokenName,
+        string memory _tokenSymbol,
+        address _randomizerContract
+    ) ERC721(_tokenName, _tokenSymbol) {
+        admin = msg.sender;
+        emit PlatformUpdated(FIELD_OWNER);
+
+        isWhitelisted[msg.sender] = true;
+        emit AllowlistUpdated(msg.sender, true);
+
+        renderProviderAddress = payable(msg.sender);
+        emit PlatformUpdated(FIELD_RENDER_PROVIDER_ADDRESS);
+
+        randomizerContract = IRandomizer(_randomizerContract);
+        emit PlatformUpdated(FIELD_RANDOMIZER_ADDRESS);
```

Comment:
to follow DRY principle, we could have update of admin, isWhitelisted, renderProviderAddress, and randomizerContract all be modified + events emitted in internal functions, something like calling the following from the constructor and updateAdmin function:
```
    /**
     * @notice Updates contract admin to `_adminAddress`.
     */
    function _updateAdmin(address _adminAddress) internal {
        admin = _adminAddress;
        emit PlatformUpdated(FIELD_OWNER);
    }
    ```
    WDYT? I can update the V3 events PR to match if you want to do this 👍

Reply:
yep good call, i say lets do it 👍

Reply:
will update here with relevant changes shortly